pkg/ring/kv/multi.go

```diff
@@ -0,0 +1,335 @@
+		}
+
+		if cancelCtx.Err() == context.Canceled && origCtx.Err() == nil {
+			// our context was cancelled, but outer context is not done yet. retry
+			continue
+		}
+
+		return err
+	}
+}
+
+func (m *multiClient) Get(ctx context.Context, key string) (interface{}, error) {
+	var result interface{}
+	err := m.runWithPrimaryClient(ctx, func(newCtx context.Context, primary kvclient) error {
+		var err2 error
+		result, err2 = primary.client.Get(newCtx, key)
+		return err2
+	})
+	return result, err
+}
+
+func (m *multiClient) CAS(ctx context.Context, key string, f func(in interface{}) (out interface{}, retry bool, err error)) error {
+	return m.runWithPrimaryClient(ctx, func(newCtx context.Context, primary kvclient) error {
+		return primary.client.CAS(newCtx, key, f)
+	})
+}
+
+func (m *multiClient) WatchKey(ctx context.Context, key string, f func(interface{}) bool) {
+	_ = m.runWithPrimaryClient(ctx, func(newCtx context.Context, primary kvclient) error {
+		primary.client.WatchKey(newCtx, key, f)
+		return newCtx.Err()
+	})
+}
+
+func (m *multiClient) WatchPrefix(ctx context.Context, prefix string, f func(string, interface{}) bool) {
+	_ = m.runWithPrimaryClient(ctx, func(newCtx context.Context, primary kvclient) error {
+		primary.client.WatchPrefix(newCtx, prefix, f)
+		return newCtx.Err()
+	})
+}
+
+func (m *multiClient) watchChanges(prefix string) {
+	for m.ctx.Err() == nil {
+		err := m.runWithPrimaryClient(m.ctx, func(newCtx context.Context, primary kvclient) error {
+			primary.client.WatchPrefix(newCtx, prefix, func(key string, val interface{}) bool {
+				level.Debug(util.Logger).Log("msg", "value updated", "key", key, "store", primary.name)
+
+				if m.mirroringEnabled.Load() {
+					// don't pass new context to keyUpdated, we don't want to react on primary client changes here.
+					m.keyUpdated(primary, key, val)
+				}
+
+				err := m.rateLimiter.Wait(newCtx)
+				if err != nil && err != context.Canceled {
+					level.Error(util.Logger).Log("msg", "error while rate limiting multi-client", "err", err)
+				}
+
+				return true
+			})
+
+			return m.ctx.Err()
+		})
+
+		if err != nil {
+			level.Warn(util.Logger).Log("msg", "watching changes returned error", "err", err)
+		}
+	}
+}
+
+func (m *multiClient) keyUpdated(primary kvclient, key string, newValue interface{}) {
+	// let's propagate this to all remaining clients
+	for _, kvc := range m.clients {
+		if kvc == primary {
+			continue
+		}
+
+		stored := true
+		err := kvc.client.CAS(m.ctx, key, func(in interface{}) (out interface{}, retry bool, err error) {
+			stored = true
+
+			if eq, ok := in.(withEqual); ok && eq.Equal(newValue) {
+				stored = false
+				level.Debug(util.Logger).Log("msg", "no change", "key", key)
+				return nil, false, nil
+			}
+
+			// try once
+			return in, false, nil
+		})
+
+		if err != nil {
+			level.Error(util.Logger).Log("msg", "failed to update value in secondary store", "key", key, "err", err, "primary", primary.name, "secondary", kvc.name)
```

Comment:
If we fail to update here does that mean going forward it will break the migration to the new ring? Is their any redundancies or does this mean we need to start fresh and roll out a new set of ingesters? That should be fine since this is a rare procedure just good to know.

Reply:
If this fails, something bad happened, e.g. Consul/Etcd is down, or CAS operation failed too many times. However, watching still continues, because WatchPrefix only ends when either context is done, or 'f' returns false (which we don't do in watchChanges).

As for redundancies – idea is that ALL clients accessing the ring will be configured to use multiClient and mirroring at the same time, so if at least one is still alive, mirroring is done. This is also reason why we only try to update the value once.

Reply:
👍